### PR TITLE
[SPARK-47603][KUBERNETES][YARN] Resource managers: Migrate logWarn with variables to structured logging framework

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to migrate `logWanrning` in module `Resource managers` with variables to `structured logging framework`.

### Why are the changes needed?
To enhance Apache Spark's logging system by implementing structured logging.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala [hunk 1]
```diff
@@ -24,7 +24,8 @@ import org.apache.spark.deploy.k8s._
 import org.apache.spark.deploy.k8s.Config._
 import org.apache.spark.deploy.k8s.Constants._
 import org.apache.spark.deploy.k8s.submit._
-import org.apache.spark.internal.Logging
+import org.apache.spark.internal.{Logging, MDC}
+import org.apache.spark.internal.LogKey.{CONFIG, CONFIG2, CONFIG3, CONFIG4, CONFIG5}
```
Comment (panbingkun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala:28]: Although this looks `a bit ugly`, there seems to be no better way.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 2]
```diff
@@ -857,7 +858,7 @@ private[spark] class ApplicationMaster(
             finish(FinalApplicationStatus.SUCCEEDED, ApplicationMaster.EXIT_SUCCESS)
           } else {
             logError(log"Driver terminated with exit code ${MDC(EXIT_CODE, exitCode)}! " +
-              log"Shutting down. ${MDC(REMOTE_ADDRESS, remoteAddress)}")
+              log"Shutting down. ${MDC(RPC_ADDRESS, remoteAddress)}")
```
Comment (panbingkun) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:861]: This seems unrelated to this PR, but I found that the LogKey `REMOTE_ADDRESS` and `RPC_ADDRESS` have the same meaning, so I removed `REMOTE_ADDRESS` from `LogKey` and only kept `RPC_ADDRESS`

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 3]
```diff
@@ -110,7 +121,8 @@ object LogKey extends Enumeration {
   val REDUCE_ID = Value
   val RELATION_NAME = Value
   val REMAINING_PARTITIONS = Value
-  val REMOTE_ADDRESS = Value
```
Comment (panbingkun) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:113]: I found that the LogKey `REMOTE_ADDRESS` and `RPC_ADDRESS` have the same meaning, so I removed `REMOTE_ADDRESS` from `LogKey` and only kept `RPC_ADDRESS`

**Comment by panbingkun** [comment]
cc @gengliangwang

File: common/utils/src/main/scala/org/apache/spark/internal/Logging.scala [hunk 4]
```diff
@@ -105,9 +105,10 @@ trait Logging {
       val context = new java.util.HashMap[String, String]()
 
       args.foreach { mdc =>
-        sb.append(mdc.value.toString)
+        val value = if (mdc.value != null) mdc.value.toString else null
```
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/Logging.scala:108]: let's rebase the code.

Comment (panbingkun) [reply common/utils/src/main/scala/org/apache/spark/internal/Logging.scala:108]: Okay

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 5]
```diff
@@ -210,10 +211,10 @@ class ExecutorPodsAllocator(
     }
 
     if (timedOut.nonEmpty) {
-      logWarning(s"Executors with ids ${timedOut.mkString(",")} were not detected in the" +
-        s" Kubernetes cluster after $podCreationTimeout ms despite the fact that a previous" +
-        " allocation attempt tried to create them. The executors may have been deleted but the" +
-        " application missed the deletion event.")
+      logWarning(log"Executors with ids ${MDC(RESOURCE_NAME, timedOut.mkString(","))}} were not " +
```
Comment (gengliangwang) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:214]: Should this be EXECUTOR_IDS?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 6]
```diff
@@ -282,7 +283,7 @@ class ExecutorPodsAllocator(
 
       val newFailedExecutorIds = currentFailedExecutorIds.diff(failedExecutorIds)
       if (newFailedExecutorIds.nonEmpty) {
-        logWarning(s"${newFailedExecutorIds.size} new failed executors.")
+        logWarning(log"${MDC(RESOURCE_COUNT, newFailedExecutorIds.size)} new failed executors.")
```
Comment (gengliangwang) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:286]: Let's make it simply "COUNT"

**Comment by gengliangwang** [comment]
@panbingkun Thanks for the work. LGTM except for two comments.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala (hunk 5 above)
Comment (panbingkun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:214]: Oh, yes, I misread the `position` of the `tuple`, updated

**Comment by panbingkun** [comment]
> @panbingkun Thanks for the work. LGTM except for two comments.

Updated, done.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Pull request closed without merging**